In [1]:
!pip install hmmlearn

Defaulting to user installation because normal site-packages is not writeable
  Obtaining dependency information for hmmlearn from https://files.pythonhosted.org/packages/c9/50/51a423eb735cfca35ef9b6b9f8155936f7843ac32f481ce9b79bf7e9cab5/hmmlearn-0.3.3-cp39-cp39-macosx_10_9_universal2.whl.metadata
  Obtaining dependency information for numpy>=1.10 from https://files.pythonhosted.org/packages/96/ff/06d1aa3eeb1c614eda245c1ba4fb88c483bee6520d361641331872ac4b82/numpy-2.0.2-cp39-cp39-macosx_14_0_arm64.whl.metadata
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.9/60.9 kB 2.5 MB/s eta 0:00:00
  Obtaining dependency information for scikit-learn!=0.22.0,>=0.16 from https://files.pythonhosted.org/packages/1b/be/386ef63d9d5e2ddf8308f6a164e4b388d5c5aecc0504d25acc6b33d8b09e/scikit_learn-1.5.2-cp39-cp39-macosx_12_0_arm64.whl.metadata
  Obtaining dependency information for scipy>=0.19 from https://files.pythonhosted.org/packages/5c/c0/e71b94b20ccf9effb38d7147c0064c08c622309fd487b1b677771a97d18c/sci

In [12]:
import math
import numpy as np
from hmmlearn import hmm

states = ('Rainy', 'Sunny')
observations = ('walk', 'shop', 'clean')
 
start_probability = {'Rainy': 0.6, 'Sunny': 0.4}
 
transition_probability = {
   'Rainy' : {'Rainy': 0.7, 'Sunny': 0.3},
   'Sunny' : {'Rainy': 0.4, 'Sunny': 0.6},
   }
 
emission_probability = {
   'Rainy' : {'walk': 0.1, 'shop': 0.4, 'clean': 0.5},
   'Sunny' : {'walk': 0.6, 'shop': 0.3, 'clean': 0.1},
   }

model = hmm.CategoricalHMM(n_components=2)
model.startprob_ = np.array([0.6, 0.4])
model.transmat_ = np.array([[0.7, 0.3],
                            [0.4, 0.6]])
model.emissionprob_ = np.array([[0.1, 0.4, 0.5],
                                [0.6, 0.3, 0.1]])

Problem 1: Given a known model what is the likelihood of sequence O happening?

Answer: The probability of the first observation being “walk” equals to the multiplication of the initial state distribution and emission probability matrix. 0.6 x 0.1 + 0.4 x 0.6 = 0.30 (30%). The log likelihood is provided from calling .score.

In [13]:
walk = 0; shop = 1; clean = 2

p1 = math.exp(model.score(np.array([[walk]])))

p2 = math.exp(model.score(np.array([[shop]])))

p3 = math.exp(model.score(np.array([[clean]])))

p4 = math.exp(model.score(np.array([[clean,clean,clean]])))

print(p1,p2,p3,p4)

0.30000000000000004 0.36000000000000004 0.3400000000000001 0.04590400000000001


Problem 2: Given a known model and sequence O, what is the optimal hidden state sequence? This will be useful if we want to know if the weather is “Rainy” or “Sunny”.

In [ ]:
logprob, seq = model.decode(np.array([[shop,clean,walk]]).transpose())
print(math.exp(logprob))
print(['Rainy' if x==0 else 'Sunny' for x in seq])

logprob, seq = model.decode(np.array([[clean,clean,clean]]).transpose())
print(math.exp(logprob))
print(['Rainy' if x==0 else 'Sunny' for x in seq])

0.015120000000000003
['Rainy', 'Rainy', 'Sunny']
0.03674999999999999
['Rainy', 'Rainy', 'Rainy']


- Given the known model and the observation {“Shop”, “Clean”, “Walk”}, the weather was most likely {“Rainy”, “Rainy”, “Sunny”} with ~1.5% probability.
- Given the known model and the observation {“Clean”, “Clean”, “Clean”}, the weather was most likely {“Rainy”, “Rainy”, “Rainy”} with ~3.6% probability.